tools/kompos/src/messages.ts

```diff
@@ -115,12 +115,137 @@ export interface SectionBreakpointData extends AbstractBreakpointData {
   coord: FullSourceCoordinate;
 }
 
-export function createLineBreakpointData(sourceUri: string, line: number): LineBreakpointData {
-  return {
-    type: "LineBreakpoint",
-    line:      line,
-    sourceUri: sourceUri,
-    enabled:   false};
+export type Breakpoint = LineBreakpoint | MessageBreakpoint |
+  AsyncMethodRcvBreakpoint | PromiseBreakpoint;
+
+abstract class AbstractBreakpoint<T extends AbstractBreakpointData> {
+  readonly data: T;
+  checkbox: any;
+  readonly source: Source;
+
+  constructor(data: T, source: Source) {
+    this.data     = data;
+    this.checkbox = null;
+    this.source   = source;
+  }
+
+  /**
+   * @return a unique id for the breakpoint, to be used in the view as HTML id
+   */
+  getId() {
+    return 'bp:';
+  }
+
+  toggle() {
+    this.data.enabled = !this.data.enabled;
+  }
+
+  isEnabled() {
+    return this.data.enabled;
+  }
+}
+
+export class LineBreakpoint extends AbstractBreakpoint<LineBreakpointData> {
+  readonly lineNumSpan: Element;
+  readonly sourceId: string;
+
+  constructor(data: LineBreakpointData, source: Source, sourceId: string,
+      lineNumSpan: Element) {
+    super(data, source);
+    this.lineNumSpan = lineNumSpan;
+    this.sourceId    = sourceId;
+  }
+
+  getId(): string {
+    return super.getId() + this.sourceId + ':' + this.data.line;
+  }
+}
+
+export class MessageBreakpoint extends AbstractBreakpoint<SectionBreakpointData> {
+  readonly sectionId: string;
+
+  constructor(data: SectionBreakpointData, source: Source, sectionId: string) {
+    super(data, source);
+    this.sectionId = sectionId;
+  }
+
+  getId(): string {
+    return super.getId() + this.sectionId; /** why can not add :message in the string identification of the breakpoint? */
+  }
+}
+
+export class AsyncMethodRcvBreakpoint extends AbstractBreakpoint<SectionBreakpointData> {
+  readonly sectionId: string;
+
+  constructor(data: SectionBreakpointData, source: Source, sectionId: string) {
+    super(data, source);
+    this.sectionId = sectionId;
+  }
+
+  getId(): string {
+    return super.getId() + this.sectionId + ":async-rcv";
+  }
+}
+
+export class PromiseBreakpoint extends AbstractBreakpoint<SectionBreakpointData> {
+  readonly sectionId: string;
+
+  constructor(data: SectionBreakpointData, source: Source, sectionId: string) {
+    super(data, source);
+    this.sectionId = sectionId;
+  }
+
+  getId(): string {
+    return super.getId() + this.sectionId + ":promise";
+  }
+}
+
+export function createLineBreakpoint(source: Source, sourceId: string,
+    line: number, clickedSpan: Element) {
+  return new LineBreakpoint({
+    type: "LineBreakpoint", line: line, sourceUri: source.uri, enabled: false},
+    source, sourceId, clickedSpan);
+}
+
+export function createMsgBreakpoint(source: Source,
```

Comment:
If the only difference between the various breakpoints is here the type. Please factor out this code to create the object that is passed to the constructor into a helper method to avoid code duplication.